problem-gson/src/main/java/org/zalando/problem/DefaultProblemAdapter.java

```diff
@@ -0,0 +1,106 @@
+package org.zalando.problem;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.TypeAdapter;
+import com.google.gson.internal.bind.TypeAdapters;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.gson.internal.bind.TypeAdapters.STRING;
+
+final class DefaultProblemAdapter extends TypeAdapter<DefaultProblem> {
+
+    private static final TypeToken<Map<String, Object>> MAP_TYPE = new TypeToken<Map<String, Object>>() {
+    };
+    private static final TypeAdapter<URI> TYPE_ADAPTER = TypeURIAdapter.INSTANCE;
+
+    private final Gson gson;
+    private final boolean stacktrace;
+    private final TypeAdapter<Map<String, Object>> mapAdapter;
+    private final TypeAdapter<StatusType> statusAdapter;
+    private final TypeAdapter<ThrowableProblem> throwableProblemAdapter;
+
+    DefaultProblemAdapter(Gson gson, boolean stacktrace) {
+        this.gson = gson;
+        this.stacktrace = stacktrace;
+        this.mapAdapter = gson.getAdapter(MAP_TYPE);
+        this.statusAdapter = gson.getAdapter(StatusType.class);
+        this.throwableProblemAdapter = gson.getAdapter(ThrowableProblem.class).nullSafe();
+    }
+
+    @Override
+    public void write(JsonWriter out, DefaultProblem problem) throws IOException {
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.add("type", TYPE_ADAPTER.toJsonTree(problem.getType()));
+        jsonObject.add("title", STRING.toJsonTree(problem.getTitle()));
+        jsonObject.add("status", statusAdapter.toJsonTree(problem.getStatus()));
+        jsonObject.add("detail", STRING.toJsonTree(problem.getDetail()));
+        jsonObject.add("instance", TypeAdapters.URI.toJsonTree(problem.getInstance()));
+        jsonObject.add("cause", throwableProblemAdapter.toJsonTree(problem.getCause()));
+        JsonObject mapObject = mapAdapter.toJsonTree(problem.getParameters()).getAsJsonObject();
+        Set<Map.Entry<String, JsonElement>> entries = mapObject.entrySet();
+        for (Map.Entry<String, JsonElement> entry : entries) {
+            jsonObject.add(entry.getKey(), entry.getValue());
+        }
+
+        if (stacktrace) {
```

Comment:
Does it correctly remove the default "stackTrace" property?

Reply:
Yes, this adapter is specifically written for `DefaultProblem` and does not delegate any serialization to Gson. So only fields added using methods `JsonObject#add` will be present in the output.

Reply:
Ah, I see.